File: bindgen-tests/lib/src/trait_interfaces.rs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

//! Trait interface test.  This is the like the `callback_interfaces` module, except it uses a
//! trait that has `with_foreign` enabled (AKA a trait interface).  This means UniFFI supports both
//! Rust and foreign language implementations of the trait.

use std::sync::{
    atomic::{AtomicU32, Ordering},
    Arc,
};

use crate::callback_interfaces::CallbackInterfaceNumbers;
use crate::errors::TestError;

#[uniffi::export(with_foreign)]
pub trait TestTraitInterface: Send + Sync {
    /// No-op function, this tests if that we can make calls at all
    fn noop(&self);
    /// Get the internal value
    fn get_value(&self) -> u32;
    /// Set the internal value
    fn set_value(&self, value: u32);
    /// Method aimed at maximizing the complexity
    ///
    /// This should return an error if `numbers.a == numbers.b` otherwise it should return numbers back
    /// unchanged.
    fn throw_if_equal(
        &self,
        numbers: CallbackInterfaceNumbers,
    ) -> Result<CallbackInterfaceNumbers, TestError>;
}

#[uniffi::export]
fn invoke_test_trait_interface_noop(interface: Arc<dyn TestTraitInterface>) {
    interface.noop()
}

#[uniffi::export]
fn invoke_test_trait_interface_get_value(interface: Arc<dyn TestTraitInterface>) -> u32 {
    interface.get_value()
}

#[uniffi::export]
fn invoke_test_trait_interface_set_value(interface: Arc<dyn TestTraitInterface>, value: u32) {
    interface.set_value(value)
}

#[uniffi::export]
fn invoke_test_trait_interface_throw_if_equal(
    interface: Arc<dyn TestTraitInterface>,
    numbers: CallbackInterfaceNumbers,
) -> Result<CallbackInterfaceNumbers, TestError> {
    interface.throw_if_equal(numbers)
}

/// Create an implementation of the interface in Rust
#[uniffi::export]
fn create_test_trait_interface(value: u32) -> Arc<dyn TestTraitInterface> {
    Arc::new(TestTraitInterfaceImpl {
        value: AtomicU32::new(value),
    })
}

struct TestTraitInterfaceImpl {
    value: AtomicU32,
}

impl TestTraitInterface for TestTraitInterfaceImpl {
    fn noop(&self) {}

    fn get_value(&self) -> u32 {
        self.value.load(Ordering::Relaxed)
    }

    fn set_value(&self, value: u32) {
        self.value.store(value, Ordering::Relaxed);
    }

    fn throw_if_equal(
        &self,
        numbers: CallbackInterfaceNumbers,
    ) -> Result<CallbackInterfaceNumbers, TestError> {
        if numbers.a == numbers.b {
            Err(TestError::Failure1)
        } else {
            Ok(numbers)
        }
    }
}

/// Async version of `TestTraitInterface`
#[uniffi::export(with_foreign)]
#[async_trait::async_trait]
pub trait AsyncTestTraitInterface: Send + Sync {
    /// No-op function, this tests if that we can make calls at all
    async fn noop(&self);
    /// Get the internal value
    async fn get_value(&self) -> u32;
    /// Set the internal value
    async fn set_value(&self, value: u32);
    /// Method aimed at maximizing the complexity
    ///
    /// This should return an error if `numbers.a == numbers.b` otherwise it should return numbers back
    /// unchanged.
    async fn throw_if_equal(
        &self,
        numbers: CallbackInterfaceNumbers,
    ) -> Result<CallbackInterfaceNumbers, TestError>;
}

#[uniffi::export]
fn create_async_test_trait_interface(value: u32) -> Arc<dyn AsyncTestTraitInterface> {
    Arc::new(TestTraitInterfaceImpl {
        value: AtomicU32::new(value),
    })
}

#[uniffi::export]
async fn invoke_async_test_trait_interface_noop(interface: Arc<dyn AsyncTestTraitInterface>) {
    interface.noop().await
}

#[uniffi::export]
async fn invoke_async_test_trait_interface_get_value(
    interface: Arc<dyn AsyncTestTraitInterface>,
) -> u32 {
    interface.get_value().await
}

#[uniffi::export]
async fn invoke_async_test_trait_interface_set_value(
    interface: Arc<dyn AsyncTestTraitInterface>,
    value: u32,
) {
    interface.set_value(value).await
}

#[uniffi::export]
async fn invoke_async_test_trait_interface_throw_if_equal(
    interface: Arc<dyn AsyncTestTraitInterface>,
    numbers: CallbackInterfaceNumbers,
) -> Result<CallbackInterfaceNumbers, TestError> {
    interface.throw_if_equal(numbers).await
}

#[async_trait::async_trait]
impl AsyncTestTraitInterface for TestTraitInterfaceImpl {
    async fn noop(&self) {}

    async fn get_value(&self) -> u32 {
        self.value.load(Ordering::Relaxed)
    }

    async fn set_value(&self, value: u32) {
        self.value.store(value, Ordering::Relaxed);
    }

    async fn throw_if_equal(
        &self,
        numbers: CallbackInterfaceNumbers,
    ) -> Result<CallbackInterfaceNumbers, TestError> {
        if numbers.a == numbers.b {
            Err(TestError::Failure1)
        } else {
            Ok(numbers)
        }
    }
}

// Functions to round-trip trait interfaces across the FFI
//
// These are used to test that they still work after being lifted/lowered.
// The list versions are there to test them being passed via a `RustBuffer`.

#[uniffi::export]
fn roundtrip_test_trait_interface(
    interface: Arc<dyn TestTraitInterface>,
) -> Arc<dyn TestTraitInterface> {
    interface
}

#[uniffi::export]
fn roundtrip_test_trait_interface_list(
    interfaces: Vec<Arc<dyn TestTraitInterface>>,
) -> Vec<Arc<dyn TestTraitInterface>> {
    interfaces
}

#[uniffi::export]
fn roundtrip_async_test_trait_interface(
    interface: Arc<dyn AsyncTestTraitInterface>,
) -> Arc<dyn AsyncTestTraitInterface> {
    interface
}

#[uniffi::export]
fn roundtrip_async_test_trait_interface_list(
    interfaces: Vec<Arc<dyn AsyncTestTraitInterface>>,
) -> Vec<Arc<dyn AsyncTestTraitInterface>> {
    interfaces
}

File: bindgen-tests/swift/tests/trait_interfaces.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation
import uniffi_bindgen_tests

var dispatchGroup = DispatchGroup()
var traitRefCount = 0;

class TraitImpl: TestTraitInterface, @unchecked Sendable {
    var value: UInt32;

    init(value: UInt32) {
        self.value = value;
        traitRefCount += 1
    }

    deinit {
        traitRefCount -= 1
    }

    func noop() { }

    func getValue() -> UInt32 {
        return self.value;
    }

    func setValue(value: UInt32) {
        self.value = value;
    }

    func throwIfEqual(numbers: CallbackInterfaceNumbers) throws -> CallbackInterfaceNumbers {
        if numbers.a == numbers.b {
            throw TestError.Failure1
        }
        return numbers
    }

    func roundtripRecord(value: SimpleRec) -> SimpleRec { value }

    func roundtripEnum(value: EnumWithData) -> EnumWithData { value }

    func roundtripInterface(value: TestInterface) -> TestInterface { value }
}

// Test calling the Rust impl from Swift
func testRustImpl(traitImpl: TestTraitInterface) {
    traitImpl.noop();
    assert(traitImpl.getValue() == 42)
    traitImpl.setValue(value: 43);
    assert(traitImpl.getValue() == 43)
    do {
        let _ = try traitImpl.throwIfEqual(numbers: CallbackInterfaceNumbers(a: 10, b: 10))
        fatalError("expected throwIfEqual to throw")
    } catch TestError.Failure1 {
        // expected
    } catch {
        fatalError("unexpected error \(error)")
    }

    assert(
        try! traitImpl.throwIfEqual(numbers: CallbackInterfaceNumbers(a: 10, b: 11)) ==
        CallbackInterfaceNumbers(a: 10, b: 11))

    assert(traitImpl.roundtripRecord(value: SimpleRec(a: 10)) == SimpleRec(a: 10))
    assert(
        traitImpl.roundtripEnum(value: EnumWithData.a(value: 10, value2: 20)) ==
        EnumWithData.a(value: 10, value2: 20))
    assert(traitImpl.roundtripInterface(value: TestInterface(value: 20)).getValue() == 20)
}

testRustImpl(traitImpl: createTestTraitInterface(value: 42))
testRustImpl(traitImpl: roundtripTestTraitInterface(interface: createTestTraitInterface(value: 42)))
testRustImpl(traitImpl: roundtripTestTraitInterfaceList(interfaces: [createTestTraitInterface(value: 42)])[0])

// Test calling the Swift impl by roundtripping it through a Rust function
func testSwiftImpl(traitImpl: TestTraitInterface) {
    invokeTestTraitInterfaceNoop(interface: traitImpl)
    assert(invokeTestTraitInterfaceGetValue(interface: traitImpl) == 42)
    invokeTestTraitInterfaceSetValue(interface: traitImpl, value: 43)
    assert(invokeTestTraitInterfaceGetValue(interface: traitImpl) == 43)
    do {
        let _ = try invokeTestTraitInterfaceThrowIfEqual(
            interface: traitImpl,
            numbers: CallbackInterfaceNumbers(a: 10, b: 10)
        )
        fatalError("expected throwIfEqual to throw")
    } catch TestError.Failure1 {
        // expected
    } catch {
        fatalError("unexpected error \(error)")
    }

    assert(
        try! invokeTestTraitInterfaceThrowIfEqual(
            interface: traitImpl,
            numbers: CallbackInterfaceNumbers(a: 10, b: 11)
        ) == CallbackInterfaceNumbers(a: 10, b: 11))

    assert(
        invokeTestTraitInterfaceRoundtripRecord(interface: traitImpl, rec: SimpleRec(a: 10)) ==
        SimpleRec(a: 10))
    assert(
        invokeTestTraitInterfaceRoundtripEnum(interface: traitImpl, en: EnumWithData.a(value: 10, value2: 20)) ==
        EnumWithData.a(value: 10, value2: 20))
    assert(invokeTestTraitInterfaceRoundtripInterface(interface: traitImpl, iface: TestInterface(value: 20)).getValue() == 20)
}

testSwiftImpl(traitImpl: TraitImpl(value: 42))
testSwiftImpl(traitImpl: roundtripTestTraitInterface(interface: TraitImpl(value: 42)))
testSwiftImpl(traitImpl: roundtripTestTraitInterfaceList(interfaces: [TraitImpl(value: 42)])[0])

var asyncTraitRefCount = 0;

class AsyncTraitImpl: AsyncTestTraitInterface, @unchecked Sendable {
    var value: UInt32;

    init(value: UInt32) {
        self.value = value;
        asyncTraitRefCount += 1
    }

    deinit {
        asyncTraitRefCount -= 1
    }

    func noop() async { }

    func getValue() async -> UInt32 {
        return self.value;
    }

    func setValue(value: UInt32) async {
        self.value = value;
    }

    func throwIfEqual(numbers: CallbackInterfaceNumbers) async throws -> CallbackInterfaceNumbers {
        if numbers.a == numbers.b {
            throw TestError.Failure1
        }
        return numbers
    }
}

func testAsyncRustImpl(traitImpl: AsyncTestTraitInterface) async {
    await traitImpl.noop();
    var valueReturn = await traitImpl.getValue();
    assert(valueReturn == 42)
        await traitImpl.setValue(value: 43);
    valueReturn = await traitImpl.getValue();
    assert(valueReturn == 43)
        do {
            let _ = try await traitImpl.throwIfEqual(numbers: CallbackInterfaceNumbers(a: 10, b: 10))
                fatalError("expected throwIfEqual to throw")
        } catch TestError.Failure1 {
            // expected
        } catch {
            fatalError("unexpected error \(error)")
        }

    let numbersReturn = try! await traitImpl.throwIfEqual(numbers: CallbackInterfaceNumbers(a: 10, b: 11))
    assert(numbersReturn == CallbackInterfaceNumbers(a: 10, b: 11))
}

dispatchGroup.enter()
Task {
    await testAsyncRustImpl(traitImpl: createAsyncTestTraitInterface(value: 42))
    await testAsyncRustImpl(traitImpl: roundtripAsyncTestTraitInterface(interface: createAsyncTestTraitInterface(value: 42)))
    await testAsyncRustImpl(traitImpl: roundtripAsyncTestTraitInterfaceList(interfaces: [createAsyncTestTraitInterface(value: 42)])[0])
    dispatchGroup.leave()
}
dispatchGroup.wait()

func testAsyncSwiftImpl(traitImpl: AsyncTestTraitInterface) async {
    await invokeAsyncTestTraitInterfaceNoop(interface: traitImpl)
    var val = await invokeAsyncTestTraitInterfaceGetValue(interface: traitImpl)
    assert(val == 42)
    await invokeAsyncTestTraitInterfaceSetValue(interface: traitImpl, value: 43)
    val = await invokeAsyncTestTraitInterfaceGetValue(interface: traitImpl)
    assert(val == 43)
    do {
        let _ = try await invokeAsyncTestTraitInterfaceThrowIfEqual(
            interface: traitImpl,
            numbers: CallbackInterfaceNumbers(a: 10, b: 10)
        )
        fatalError("expected throwIfEqual to throw")
    } catch TestError.Failure1 {
        // expected
    } catch {
        fatalError("unexpected error \(error)")
    }

    let numbersReturn = try! await invokeAsyncTestTraitInterfaceThrowIfEqual(
        interface: traitImpl,
        numbers: CallbackInterfaceNumbers(a: 10, b: 11)
    )
    assert(numbersReturn == CallbackInterfaceNumbers(a: 10, b: 11))
}

dispatchGroup.enter()
Task {
    await testAsyncSwiftImpl(traitImpl: AsyncTraitImpl(value: 42))
    await testAsyncSwiftImpl(traitImpl: roundtripAsyncTestTraitInterface(interface: AsyncTraitImpl(value: 42)))
    await testAsyncSwiftImpl(traitImpl: roundtripAsyncTestTraitInterfaceList(interfaces: [AsyncTraitImpl(value: 42)])[0])
    dispatchGroup.leave()
}
dispatchGroup.wait()

// The previous calls created and destroyed a ton of references to the Swift-implemented trait
// interfaces, check that the refcounts have gone back to 0.
assert(traitRefCount == 0)
assert(asyncTraitRefCount == 0)

// Rust-only trait: smoke-test for the #[uniffi::export(rust)] codegen path
func testRustOnlyImpl(impl: TestRustOnlyTraitInterface) {
    do {
        let _ = try impl.throwIfEqual(numbers: CallbackInterfaceNumbers(a: 10, b: 10))
        fatalError("expected throwIfEqual to throw")
    } catch TestError.Failure1 {
        // expected
    } catch {
        fatalError("unexpected error \(error)")
    }
    assert(
        try! impl.throwIfEqual(numbers: CallbackInterfaceNumbers(a: 10, b: 11)) == CallbackInterfaceNumbers(a: 10, b: 11))
}

testRustOnlyImpl(impl: createRustOnlyTestTraitInterface())

// Foreign-only trait: smoke-test for the #[uniffi::export(foreign)] codegen path
class ForeignOnlyImpl: TestForeignOnlyTraitInterface, @unchecked Sendable {
    func throwIfEqual(numbers: CallbackInterfaceNumbers) throws -> CallbackInterfaceNumbers {
        if numbers.a == numbers.b {
            throw TestError.Failure1
        }
        return numbers
    }
}

func testForeignOnlyImpl(impl: TestForeignOnlyTraitInterface) {
    do {
        let _ = try invokeTestForeignOnlyTraitThrowIfEqual(
            interface: impl,
            numbers: CallbackInterfaceNumbers(a: 10, b: 10)
        )
        fatalError("expected throwIfEqual to throw")
    } catch TestError.Failure1 {
        // expected
    } catch {
        fatalError("unexpected error \(error)")
    }
    assert(
        try! invokeTestForeignOnlyTraitThrowIfEqual(
            interface: impl,
            numbers: CallbackInterfaceNumbers(a: 10, b: 11)
        ) == CallbackInterfaceNumbers(a: 10, b: 11))
}

testForeignOnlyImpl(impl: ForeignOnlyImpl())
testForeignOnlyImpl(impl: roundtripTestForeignOnlyTrait(interface: ForeignOnlyImpl()))
